Adding ProjectClass.cs to DistroRickDev/PMS-Core Repository

ivanmartin91 opened the pull request:
ProjectClass is based on the group project instructions and class diagram, it shows the connection between the project and tasks considering each project has a list of tasks.

ivanmartin91 committed:
Adding ProjectClass.cs, updating with the agreed commenting format,
addToReport and doing a separate class for its testing

Combined 6 commits

DistroRickDev commented on PMSCore/ProjectClass.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace PMSCore {
+    //class Task with attributes: Name, Description, Status; private attribute Status demonstrates encapsulation
+    public class Task { 
+        public string Name { get; set; } //naming the Task
+        public string Description { get; set; } //providing a brief description of the Task
+        public string Status { get; private set; } = "ToDo"; //initial Status is ToDo, it is protected (only accessible from within the class and derived classes)
+
+        public Task(string name, string description) //constructor part, to initialize a new task, hiding internal components of the Task => abstraction   
+        {
+            Name = name; //setting the name for it
+            Description = description; //providing a description
+        }
+
+        public void UpdateStatus(string newStatus) //method that updates a new value for this task => abstraction (status can't be changed directly)
+        {
+            if (newStatus == "ToDo" || newStatus == "In Progress" || newStatus == "Completed") //update status only if it matches one of these: "ToDo", "In Progress", "Completed"
+                Status = newStatus;
+        }
+    }
```
This is not part of this ticket nor this PR there is work in progress for the TaskClass, if you need it you should create an empty one

DistroRickDev commented on PMSCore/ProjectClass.cs:
```diff
@@ -0,0 +1,50 @@
+                Status = newStatus;
+        }
+    }
+    //class Project which is a group of tasks, also its attributes and methods are another example of Abstraction
```
Please let's keep the class comment documentation unifrom please see this example:
https://github.com/DistroRickDev/PMS-Core/pull/7/files
Also no need to inline comment every code line the code speaks for itself

DistroRickDev commented on PMSCore/ProjectClass.cs:
```diff
@@ -0,0 +1,50 @@
+        public string Name { get; set; } //attribute to give this project a name
+        public string Description { get; set; } //providing the project's description
+        public DateTime CreationDate { get; } = DateTime.Now; //creation date, according to the instructions that each project will have a timestamp of its creation
+        public DateTime Deadline { get; set; } //instructions state that the project needs to have a deadline in addition to a name and description
```
What about project Status? 
Also suggest creating an Enum called ProjectStatus for it

DistroRickDev commented on PMSCore/ProjectClass.cs:
```diff
@@ -0,0 +1,50 @@
+    public class Project { 
+        public string Name { get; set; } //attribute to give this project a name
+        public string Description { get; set; } //providing the project's description
+        public DateTime CreationDate { get; } = DateTime.Now; //creation date, according to the instructions that each project will have a timestamp of its creation
```
this should be initialized like this in the costructor I would say

DistroRickDev reviewed (commented):
Looks solid but other then the comments left I have a couple aspects I would like you to consider:

- Class you created should be unit tested in this PR
- Project should have a status.
- Maybe an internal string that appends every action as a log (so in the end a report of the project can be generated) what do you think

ivanmartin91 commented:
@DistroRickDev , just updated the ProjectClass.cs based on your feedback, please have a look when you get a chance. Thanks!

DistroRickDev commented:
> @DistroRickDev , just updated the ProjectClass.cs based on your feedback, please have a look when you get a chance. Thanks!

@ivanmartin91  The project class was ok you need to add a new file in the test Directory to test for the class created as done here: https://github.com/DistroRickDev/PMS-Core/pull/8

ivanmartin91 commented:
@DistroRickDev , adding a new file in PMSCore.Test for the class unit tests and also reverting the code to the version without unit tests in PMSCore, hope this is okay now.

DistroRickDev commented on PMSCore/ProjectClass.cs:
```diff
@@ -0,0 +1,73 @@
+        /// </summary>
+        public void GenerateReport()
+        {
+            Console.WriteLine("Action Log:");
+            Console.WriteLine(actionLog);
```
maybe have something like this instead:
```c#
private string report;
private void addToReport(string message)
{
   report += $"[ActionLog] ${DateTime.Now().ToString()} : ${message}"
}
public GenerateReport() 
{
   return _report;
}
```

DistroRickDev commented on PMSCore/ProjectClass.cs:
```diff
@@ -0,0 +1,73 @@
+            Console.WriteLine(actionLog);
+        }
+    }
+
+    /// <summary>
+    /// Represents a task to be defined later.
+    /// </summary>
+    public class Task { }
```
SOLID principles, singe responsability purpose each class should have one responsability, please movr this to Task.cs (every file is treated as class)

DistroRickDev reviewed (commented):
Great work only a couple of final suggestions, and please squash the commits

ivanmartin91 commented:
Hey @DistroRickDev , just removed those 4 lines and put them as a separate Task.cs, also added addToReport private string report Field and methods addToReport() and GenerateReport(). Finally tried to squash those two commits but it didn't work: tried by installing GitHub desktop, using Terminal and also from the Git Repository section by right clicking but ''squash commits'' is greyed out (see attached).
![11](https://github.com/user-attachments/assets/3d378d0b-8614-4fad-8bca-875d9235dd0c)

DistroRickDev commented:
Looks good to me no apparent issues found

DistroRickDev reviewed (approved):
LGTM
